### PR TITLE
Improve tests which train model

### LINKED ISSUE #8117: Improve tests which train model
**Description of Problem**:
[We identified a couple of tests which train models in the CI and which can be improved](https://www.notion.so/rasa/Proposal-for-tests-which-train-models-a091d698ef794b1fb8f0f89c752f7907?d=fb0dc5b6-482c-4e94-9c97-199bc46bc429#963b3348c938429b9a204b012deae40b). This is supposed to speed up the CI as model trainings take quite some time.

**Overview of the Solution**:
Improve these test cases according to the suggestions. If their is no good solution attach a `nofix` label

**Examples** (if relevant):

**Blockers** (if relevant):
Should be done after

**Definition of Done**:
- [ ] Implement solutions for all issues in the `needs tackling / can be tackled` column which do not have a `nofix` label.

### PR DESCRIPTION
**Pull request by alwx**
**Proposed changes**:
- fix #8117

**Status (please check what you already did)**:
- [ ] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [ ] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

File: tests/core/test_agent.py [hunk 1]
```diff
@@ -74,6 +74,27 @@ async def test_training_data_is_reproducible():
         assert str(x.as_dialogue()) == str(same_training_data[i].as_dialogue())
 
 
+async def test_agent_train(trained_rasa_model: Text):
+    moodbot_domain = Domain.load("data/test_domains/default_with_slots.yml")
```
Comment (wochinge) [inline tests/core/test_agent.py:78]: as it's not moodbot anymore:
```suggestion
    domain = Domain.load("data/test_domains/default_with_slots.yml")
```

File: tests/core/test_training.py [hunk 2]
```diff
@@ -56,25 +58,19 @@ async def test_story_visualization_with_merging(default_domain: Domain):
...
-        additional_arguments={},
-    )
-    assert True
+async def test_training_script_without_max_history_set(
+    tmp_path: Path, monkeypatch: MonkeyPatch
+):
+    tmpdir = str(tmp_path)
 
+    policy_train = Mock()
+    monkeypatch.setattr(TEDPolicy, "train", policy_train)
```
Comment (wochinge) [inline tests/core/test_training.py:67]: why not removing `TEDPolicy` from the config file in the first place instead of mocking?

File: tests/core/test_training.py [hunk 3]
```diff
@@ -56,25 +58,19 @@ async def test_story_visualization_with_merging(default_domain: Domain):
...
 
-async def test_training_script_without_max_history_set(tmp_path: Path):
-    tmpdir = str(tmp_path)
     await train(
         DEFAULT_DOMAIN_PATH_WITH_SLOTS,
         DEFAULT_STORIES_FILE,
         tmpdir,
-        interpreter=RegexInterpreter(),
+        interpreter=Mock(spec=RasaNLUInterpreter),
```
Comment (wochinge) [inline tests/core/test_training.py:73]: is that really a performance difference?

File: tests/core/test_training.py [hunk 4]
```diff
@@ -122,31 +123,34 @@ async def test_training_script_with_restart_stories(tmp_path: Path):
...
+async def test_random_seed(tmp_path: Path, monkeypatch: MonkeyPatch):
+    policies_config = {
+        "policies": [
+            {"name": TEDPolicy.__name__, "random_seed": 42},
+            {"name": RulePolicy.__name__},
+        ]
+    }
 
+    policy_train = Mock()
+    monkeypatch.setattr(TEDPolicy, "train", policy_train)
```
Comment (wochinge) [inline tests/core/test_training.py:135]: mhm, if we mock this, then the test doesn't test anymore what it's supposed to test, no?

File: tests/integration_tests/nlu/test_train.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
+import pytest
```
Comment (wochinge) [inline tests/integration_tests/nlu/test_train.py:1]: sorry, this changed a bit when doing the test suite split. Can we keep it in the unit tests but make sure it's part of `category_full_model_training` in [here](https://github.com/RasaHQ/rasa/blob/d862768082a7d582da13f8f88bdf552c414defa5/tests/conftest.py#L81)?

File: tests/nlu/classifiers/test_diet_classifier.py [hunk 6]
```diff
@@ -511,15 +512,25 @@ async def test_train_tensorboard_logging(
...
                     "name": "DIETClassifier",
                     EPOCHS: 1,
                     TENSORBOARD_LOG_LEVEL: log_level,
                     TENSORBOARD_LOG_DIR: str(tensorboard_log_dir),
+                    MODEL_CONFIDENCE: "linear_norm",
+                    CONSTRAIN_SIMILARITIES: True,
                     EVAL_NUM_EXAMPLES: 15,
                     EVAL_NUM_EPOCHS: 1,
                 },
+                {"name": "EntitySynonymMapper"},
```
Comment (wochinge) [inline tests/nlu/classifiers/test_diet_classifier.py:533]: we can skip this component, no? It doesn't add anything to the test

File: tests/nlu/classifiers/test_diet_classifier.py [hunk 7]
```diff
@@ -528,7 +539,7 @@ async def test_train_tensorboard_logging(
     await train(
         _config,
         path=str(tmpdir),
-        data="data/examples/rasa/demo-rasa-multi-intent.yml",
+        data="examples/moodbot/data/nlu.yml",
```
Comment (wochinge) [inline tests/nlu/classifiers/test_diet_classifier.py:542]: can we use some NLU file which isn't part of `examples/`?

File: tests/nlu/selectors/test_selectors.py [hunk 8]
```diff
@@ -279,17 +285,17 @@ async def _train_persist_load_with_different_settings(
 
 
 @pytest.mark.skip_on_windows
-async def test_train_persist_load(component_builder: ComponentBuilder, tmpdir: Path):
+@pytest.mark.parametrize("should_finetune", [True, False])
+async def test_train_persist_load(
```
Comment (wochinge) [inline tests/nlu/selectors/test_selectors.py:289]: sorry if I've stated that wrong in the Notion document, but I think these tests might have to be one after another so that the 2nd train is actually a training which finetunes 😬

File: tests/nlu/test_evaluation.py [hunk 9]
```diff
@@ -394,28 +391,26 @@ async def test_eval_data(
...
-    (_, _, persisted_path) = await train(
-        _config,
-        path=str(tmp_path),
-        data=data_importer,
-        component_builder=component_builder,
-        persist_nlu_training_data=True,
-    )
-
-    interpreter = Interpreter.load(persisted_path, component_builder)
+    with rasa.model.unpack_model(trained_simple_rasa_model) as unpacked_model_directory:
```
Comment (wochinge) [inline tests/nlu/test_evaluation.py:394]: how about having this as session wide fixture? Then we don't have to reload the model

File: tests/nlu/test_evaluation.py [hunk 10]
```diff
@@ -394,28 +391,26 @@ async def test_eval_data(
...
     assert len(intent_results) == 46
-    assert len(response_selection_results) == 46
+    assert len(response_selection_results) == 0
     assert len(entity_results) == 46
 
 
 @pytest.mark.timeout(240)  # these can take a longer time than the default timeout
-def test_run_cv_evaluation(pretrained_embeddings_spacy_config: RasaNLUModelConfig):
+def test_run_cv_evaluation(
+    monkeypatch: MonkeyPatch, pretrained_embeddings_spacy_config: RasaNLUModelConfig
```
Comment (wochinge) [inline tests/nlu/test_evaluation.py:412]:
```suggestion
    pretrained_embeddings_spacy_config: RasaNLUModelConfig
```

Comment (wochinge) [inline tests/nlu/test_evaluation.py:412]: I noticed the spacy model takes quite a while to load etc. How about not using spacy here?

File: tests/nlu/test_evaluation.py [hunk 11]
```diff
@@ -476,6 +471,12 @@ def test_run_cv_evaluation_with_response_selector():
         }
     )
 
+    # mock training
+    trainer = Trainer(nlu_config)
+    trainer.pipeline = remove_pretrained_extractors(trainer.pipeline)
+    mock = Mock(return_value=Interpreter(trainer.pipeline, None))
```
Comment (wochinge) [inline tests/nlu/test_evaluation.py:477]: 😈 nice!

File: tests/nlu/test_evaluation.py [hunk 12]
```diff
@@ -949,6 +950,25 @@ async def test_nlu_comparison(tmp_path: Path):
     # combined on the same dictionary key and cannot be plotted properly
     configs = [write_file_config(config).name, write_file_config(config).name]
 
+    # mock training
+    monkeypatch.setattr(Interpreter, "load", Mock(spec=RasaNLUInterpreter))
+    monkeypatch.setattr(sys.modules["rasa.nlu"], "train", AsyncMock())
+    monkeypatch.setattr(
```
Comment (wochinge) [inline tests/nlu/test_evaluation.py:956]: I like mocking the training - I'm a bit hesitant about mocking the testing though. Are we sure the `testing` is covered by other tests?

File: tests/core/test_training.py (hunk 3 above)
Comment (alwx) [reply tests/core/test_training.py:73]: indeed, it's minimal

File: tests/nlu/classifiers/test_diet_classifier.py (hunk 6 above)
Comment (alwx) [reply tests/nlu/classifiers/test_diet_classifier.py:533]: we can. there will be a warning but we can for sure skip that

File: tests/nlu/test_evaluation.py (hunk 12 above)
Comment (alwx) [reply tests/nlu/test_evaluation.py:959]: yes, there are `test_intent_evaluation_report` and `test_intent_evaluation_report_large` so I think it's perfectly fine to mock testing for this specific test

File: tests/core/test_training.py (hunk 4 above)
Comment (alwx) [reply tests/core/test_training.py]: fixed that

File: tests/core/test_training.py [hunk 13]
```diff
@@ -121,33 +112,32 @@ async def test_training_script_with_restart_stories(tmp_path: Path, domain_path:
...
+    }
 
-@pytest.mark.parametrize("config_file", configs_for_random_seed_test())
-async def test_random_seed(
-    tmp_path: Path, config_file: Text, domain_path: Text, stories_path: Text
-):
-    # set random seed in config file to
-    # generate a reproducible training result
+    policy_train = Mock()
+    monkeypatch.setattr(TED, "fit", policy_train)
```
Comment (wochinge) [inline tests/core/test_training.py:124]: Mhm, if we don't let it `fit` then we can't really test that both prediction are the same, can we?

File: tests/nlu/classifiers/test_diet_classifier.py [hunk 14]
```diff
@@ -532,7 +542,7 @@ async def test_train_tensorboard_logging(
     await rasa.nlu.train.train(
         _config,
         path=str(tmpdir),
-        data="data/examples/rasa/demo-rasa-multi-intent.yml",
+        data="data/test_moodbot/data/nlu.yml",
```
Comment (wochinge) [inline tests/nlu/classifiers/test_diet_classifier.py:545]: How about using the `nlu_data_path` fixture?

File: tests/nlu/classifiers/test_diet_classifier.py [hunk 15]
```diff
@@ -569,7 +586,7 @@ async def test_train_model_checkpointing(
     await rasa.nlu.train.train(
         _config,
         path=str(tmpdir),
-        data="data/examples/rasa/demo-rasa.yml",
+        data="examples/moodbot/data/nlu.yml",
```
Comment (wochinge) [inline tests/nlu/classifiers/test_diet_classifier.py:589]: how about using the fixture `nlu_data_path`?

File: tests/nlu/test_evaluation.py [hunk 16]
```diff
@@ -395,37 +392,45 @@ async def test_eval_data(
...
         "data/examples/rasa/demo-rasa.json"
     )
 
+    nlu_config = RasaNLUModelConfig(
+        {
+            "language": "en",
+            "pipeline": [
+                {"name": "WhitespaceTokenizer"},
+                {"name": "CountVectorsFeaturizer"},
+                {"name": "ResponseSelector", EPOCHS: 1}
```
Comment (wochinge) [inline tests/nlu/test_evaluation.py:424]: Shouldn't this be `DIETClassifier`? How does that test even work without intent classifier 😆

Comment (wochinge) [reply tests/nlu/test_evaluation.py:424]: Can we mock the training here as well?

File: tests/core/test_training.py (hunk 13 above)
Comment (alwx) [reply tests/core/test_training.py:124]: apparently, the predictions are identical if we mock `fit` or if we don't.

this is how the predictions (`probs_1` or `probs_2`) look like in both cases:
```
{'scores': [{'action': 'action_listen', 'score': 0.0}, {'action': 'action_restart', 'score': 0.0}, {'action': 'action_session_start', 'score': 0.0}, {'action': 'action_default_fallback', 'score': 0.3}, {'action': 'action_deactivate_loop', 'score': 0.0}, {'action': 'action_revert_fallback_events', 'score': 0.0}, {'action': 'action_default_ask_affirmation', 'score': 0.0}, {'action': 'action_default_ask_rephrase', 'score': 0.0}, {'action': 'action_two_stage_fallback', 'score': 0.0}, {'action': 'action_back', 'score': 0.0}, {'action': '...', 'score': 0.0}, {'action': 'utter_channel', 'score': 0.0}, {'action': 'utter_default', 'score': 0.0}, {'action': 'utter_goodbye', 'score': 0.0}, {'action': 'utter_greet', 'score': 0.0}, {'action': 'some_form', 'score': 0.0}], 'policy': 'policy_1_RulePolicy', 'confidence': 0.3, 'tracker': {'sender_id': '1', 'slots': {'name': None, 'requested_slot': None, 'session_started_metadata': None}, 'latest_message': {'intent': {}, 'entities': [], 'text': None, 'message_id': None, 'metadata': {}}, 'latest_event_time': 1616511256.315918, 'followup_action': None, 'paused': False, 'events': [{'event': 'action', 'timestamp': 1616511256.3158872, 'name': 'action_session_start', 'policy': None, 'confidence': 1.0, 'action_text': None, 'hide_rule_turn': False}, {'event': 'session_started', 'timestamp': 1616511256.315901}, {'event': 'action', 'timestamp': 1616511256.315918, 'name': 'action_listen', 'policy': None, 'confidence': None, 'action_text': None, 'hide_rule_turn': False}], 'latest_input_channel': None, 'active_loop': {}, 'latest_action': {'action_name': 'action_listen'}, 'latest_action_name': 'action_listen'}}
```

File: tests/nlu/test_evaluation.py (hunk 16 above)
Comment (alwx) [reply tests/nlu/test_evaluation.py]: Done both!

**Review by wochinge — APPROVED** [review]
💯

File: tests/core/test_training.py (hunk 13 above)
Comment (wochinge) [reply tests/core/test_training.py]: oh dear 🤦🏻 probably becaues there is no conversation history or user message
